apps/SparseMatrix/sparse_matrix.cuh

```diff
@@ -0,0 +1,191 @@
+
+    sparse_mat_prescan<blockThreads>
+        <<<launch_box.blocks,
+           launch_box.num_threads,
+           launch_box.smem_bytes_dyn>>>(rx.get_context(), patch_ptr_v, row_ptr);
+
+    // prefix sum using CUB. 
+    void*  d_cub_temp_storage = nullptr;
+    size_t temp_storage_bytes = 0;
+    cub::DeviceScan::ExclusiveSum(
+        d_cub_temp_storage, temp_storage_bytes, row_ptr, row_ptr, num_vertices + 1);
+    CUDA_ERROR(cudaMalloc((void**)&d_cub_temp_storage, temp_storage_bytes));
+
+    cub::DeviceScan::ExclusiveSum(
+        d_cub_temp_storage, temp_storage_bytes, row_ptr, row_ptr, num_vertices + 1);
+
+    CUDA_ERROR(cudaFree(d_cub_temp_storage));
+
+    // get entry size
+    CUDA_ERROR(cudaMemcpy(&entry_size,
+                          (row_ptr + num_vertices),
+                          sizeof(uint32_t),
+                          cudaMemcpyDeviceToHost));
+
+    // printf("%" PRIu32 " - %" PRIu32 " \n", entry_size, 2 * num_edges + num_vertices);
+
+    // column index allocation and init
+    CUDA_ERROR(cudaMalloc((void**)&col_idx, entry_size * sizeof(uint32_t)));
+    rx.prepare_launch_box(
+        {Op::VV}, launch_box, (void*)sparse_mat_col_fill<blockThreads>);
+
+    sparse_mat_col_fill<blockThreads><<<launch_box.blocks,
+                                        launch_box.num_threads,
+                                        launch_box.smem_bytes_dyn>>>(
+        rx.get_context(), patch_ptr_v, row_ptr, col_idx);
+
+    // val pointer allocation, actual value init should be in another function
+    CUDA_ERROR(cudaMalloc((void**)&val, entry_size * sizeof(uint32_t)));
+}
+
+
+// TODO: add compatibility for EE, FF, VE......
+template <typename T>
+struct SparseMatInfo
+{
+    SparseMatInfo(RXMeshStatic& rx)
+        : m_patch_ptr_v(nullptr),
+          m_patch_ptr_e(nullptr),
+          m_patch_ptr_f(nullptr),
+          row_ptr(nullptr),
+          col_idx(nullptr),
+          entry_size(0),
+          val(nullptr)
+    {
+        detail::init(rx,
+                     m_patch_ptr_v,
+                     m_patch_ptr_e,
+                     m_patch_ptr_f);  // patch pointer init
+        sparse_mat_init(rx, m_patch_ptr_v, row_ptr, col_idx, entry_size, val);
+    }
+
+    void set_ones()
+    {
+        std::vector<T> init_tmp_arr(entry_size, 1);
+        CUDA_ERROR(cudaMemcpy(val,
+                              init_tmp_arr.data(),
+                              entry_size * sizeof(T),
+                              cudaMemcpyHostToDevice));
+    }
+
+    void free()
+    {
+        CUDA_ERROR(cudaFree(m_patch_ptr_v));
+        CUDA_ERROR(cudaFree(m_patch_ptr_e));
+        CUDA_ERROR(cudaFree(m_patch_ptr_f));
+        CUDA_ERROR(cudaFree(row_ptr));
+        CUDA_ERROR(cudaFree(col_idx));
+        CUDA_ERROR(cudaFree(val));
+    }
+
+    uint32_t *m_patch_ptr_v, *m_patch_ptr_e, *m_patch_ptr_f;
+    uint32_t* row_ptr;
+    uint32_t* col_idx;
+    uint32_t  entry_size;
+    T*        val;
```

Comment:
We also use this convention where class member variables are prefixed by `m_`. Since these all are device pointers, they should be also prefixed by `d_` so they should be  `m_d_col_idx`, `m_entry_size`, and `m_d_val`, `m_d_patch_ptr_v`, `m_d_patch_ptr_e`, and `m_d_patch_ptr_f`.

Reply:
done